### PR TITLE
Getting ready to start building the plan-building DSL

### PR DESCRIPTION
**Pull request by sheatrevor**

### CODE REVIEW

File: src/main/scala/com/bradbrok/filmomatic/Direction.scala [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.bradbrok.filmomatic
+
+object Direction extends Enumeration {
+  type Direction = Value
+  val In, Out = Value
```
Comment (bradbrok) [inline src/main/scala/com/bradbrok/filmomatic/Direction.scala]: Breaking it up even further sounds like a good idea. Direction, Bath, and State all are perfect.

File: src/main/scala/com/bradbrok/filmomatic/State.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+      * @example
+      * {{{
+      * assert(Idle ~> Fill)
+      * }}}
+      * @param next the subsequent value
+      * @return true if the subsequent is allowable
+      */
+    def ~>(next: State): Boolean = canBecome contains next
```
Comment (bradbrok) [inline src/main/scala/com/bradbrok/filmomatic/State.scala]: What is this Boolean statement? I'm not quite following.

File: src/main/scala/com/bradbrok/filmomatic/State.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    val canBecome = Set(AgitateA, Drain)
+  }
+  case object AgitateA extends State {
+    val direction = Some(Out)
+    val canBecome = Set(AgitateB)
+  }
+  case object AgitateB extends State {
+    val direction = Some(In)
+    val canBecome = Set(Settle, Drain)
+  }
```
Comment (bradbrok) [inline src/main/scala/com/bradbrok/filmomatic/State.scala]: Looks like we would need a "case object AgitateC" as well. I can add this.

File: src/main/scala/com/bradbrok/GPIOInterface.scala [hunk 4]
```diff
@@ -1,24 +1,6 @@
...
-import com.pi4j.io.gpio.RaspiPin;
-import com.pi4j.io.gpio.trigger.GpioCallbackTrigger;
-import com.pi4j.io.gpio.trigger.GpioPulseStateTrigger;
-import com.pi4j.io.gpio.trigger.GpioSetStateTrigger;
-import com.pi4j.io.gpio.trigger.GpioSyncStateTrigger;
-import com.pi4j.io.gpio.event.GpioPinListener;
-import com.pi4j.io.gpio.event.GpioPinDigitalStateChangeEvent;
-import com.pi4j.io.gpio.event.GpioPinEvent;
-import com.pi4j.io.gpio.event.GpioPinListenerDigital;
-import com.pi4j.io.gpio.event.PinEventType;
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/GPIOInterface.scala]: Removed unused imports.

File: src/main/scala/com/bradbrok/GPIOInterface.scala [hunk 5]
```diff
@@ -43,12 +25,12 @@ class GPIOInterface {
   val pin15 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_15, "pin15")
   val pin16 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_16, "pin16")
 
-  def setHigh(pins: List): Unit ={
+  def setHigh(pins: List[GpioPin]): Unit = {
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/GPIOInterface.scala]: Added type of the content of the list here and in the `setLow` method to get the compiler to stop complaining.

File: src/main/scala/com/bradbrok/filmomatic/Bath.scala [hunk 6]
```diff
@@ -1,6 +0,0 @@
-package com.bradbrok.filmomatic
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/Bath.scala]: Not deleted, just moved to the `com.bradbrok.filmomatic.state` package.

File: src/main/scala/com/bradbrok/filmomatic/Flow.scala [hunk 7]
```diff
@@ -1,35 +0,0 @@
-package com.bradbrok.filmomatic
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/Flow.scala]: Also not deleted, just moved to the `com.bradbrok.filmomatic.state` package.

File: src/main/scala/com/bradbrok/filmomatic/JobActor.scala [hunk 8]
```diff
@@ -1,20 +1,21 @@
...
 
 import akka.actor.Actor
-import akka.actor.Actor.Receive
 import com.bradbrok.filmomatic.JobActor.Start
 import com.bradbrok.filmomatic.Message.Request
+import com.bradbrok.filmomatic.state.Plan
 
 object JobActor {
   trait Start extends Request {
-    def state: JobState
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/JobActor.scala]: `JobState` is now called `Plan`.

File: src/main/scala/com/bradbrok/filmomatic/JobActor.scala [hunk 9]
```diff
@@ -1,20 +1,21 @@
...
+    def plan: Plan
   }
 }
 
 class JobActor extends Actor {
-  var state = JobState(1)
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/JobActor.scala]: `JobState` is now called `Plan`.

File: src/main/scala/com/bradbrok/filmomatic/JobState.scala [hunk 10]
```diff
@@ -1,13 +0,0 @@
-package com.bradbrok.filmomatic
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/JobState.scala]: Not deleted, just moved to the `com.bradbrok.filmomatic.state` package.

File: src/main/scala/com/bradbrok/filmomatic/state/Step.scala [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.bradbrok.filmomatic.state.State._
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+case class Step(state: State = Idle,
+                duration: Duration = 0 seconds,
+                delay: Duration = 0 seconds,
+                temperature: Option[Double] = None)
```
Comment (sheatrevor) [inline src/main/scala/com/bradbrok/filmomatic/state/Step.scala]: Added `temperature`, a value of `None` means the heater should be off.

File: src/test/scala/com/bradbrok/filmomatic/state/PlanSpec.scala [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        Step(Settle, duration = 5 seconds, temperature = Some(47)),
+        Step(Agitate, duration = 5 seconds, temperature = Some(47)),
+        Step(Settle, duration = 5 seconds, temperature = Some(47)),
+        Step(Agitate, duration = 5 seconds, temperature = Some(47)),
+        Step(Settle, duration = 5 seconds),
+        Step(Drain, duration = 20 seconds),
+        Step(Idle, duration = 5 seconds)
+      ))
+    ))
+    plan.isBalanced shouldBe true
```
Comment (sheatrevor) [inline src/test/scala/com/bradbrok/filmomatic/state/PlanSpec.scala]: Test which confirms that a `Plan` is 'balanced'. The definition of 'balanced' here is simply that an equal `duration` of time has been spent (in total) in both `Direction.In` and `Direction.Out`.

**Comment by sheatrevor** [comment]
Everything compiles as of right now, and the test will run using both `sbt test` and the IntelliJ ScalaTest configuration.

**Pull request merged**